### PR TITLE
Revert the PRs that made `DBNull.Value` and `NullString.Value` treated as `$null`

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

Revert the PRs that made `DBNull.Value` and `NullString.Value` treated as `$null`:
- https://github.com/PowerShell/PowerShell/pull/10781
- https://github.com/PowerShell/PowerShell/pull/9794

Also update code for Coalescing and Null-conditional operators to use 'IsNull' instead.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/LanguagePrimitives.cs [hunk 1]
```diff
@@ -593,7 +593,9 @@ public static PSDataCollection<PSObject> GetPSDataCollection(object inputValue)
         /// <param name="second">Object to compare first to.</param>
         /// <returns>True if first is equal to the second.</returns>
         public static new bool Equals(object first, object second)
-            => Equals(first, second, false, CultureInfo.InvariantCulture);
+        {
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs]: why this change?

File: src/System.Management.Automation/engine/LanguagePrimitives.cs [hunk 2]
```diff
@@ -604,7 +606,9 @@ public static new bool Equals(object first, object second)
         /// to specify the type of string comparison </param>
         /// <returns>True if first is equal to the second.</returns>
         public static bool Equals(object first, object second, bool ignoreCase)
-            => Equals(first, second, ignoreCase, CultureInfo.InvariantCulture);
+        {
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs]: same as above?

File: src/System.Management.Automation/engine/LanguagePrimitives.cs [hunk 3]
```diff
@@ -642,28 +646,25 @@ public static bool Equals(object first, object second, bool ignoreCase, IFormatP
 
             if (first == null)
             {
-                return IsNullLike(second);
+                if (second == null) return true;
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs]: could you have `return true` on new line? It is easy to misread specially because there is a return on the next line.

File: src/System.Management.Automation/engine/LanguagePrimitives.cs [hunk 4]
```diff
@@ -593,7 +593,9 @@ public static PSDataCollection<PSObject> GetPSDataCollection(object inputValue)
         /// <param name="second">Object to compare first to.</param>
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs]: @daxian-dbw, your last commit had 1 failures in `PowerShell-CI-static-analysis`
Verify Markdown Links.Verify links in /home/vsts/work/1/s/docs/debugging/README.md.https://blogs.msdn.microsoft.com/visualstudioalm/2016/03/10/experimental-net-core-debugging-in-vs-code/ should work
```powershell
retry of URL failed with error: Response status code does not indicate success: 404 (Not Found).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```

File: src/System.Management.Automation/engine/LanguagePrimitives.cs [hunk 5]
```diff
@@ -858,7 +855,7 @@ public static int Compare(object first, object second, bool ignoreCase, IFormatP
 
             // At this point, we know that they aren't equal but we have no way of
             // knowing which should compare greater than the other so we throw an exception.
-            throw PSTraceSource.NewArgumentException(nameof(first), ExtendedTypeSystem.NotIcomparable, first.ToString());
+            throw PSTraceSource.NewArgumentException("first", ExtendedTypeSystem.NotIcomparable, first.ToString());
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs]: is this change due to the revert?

**Review by adityapatwardhan — APPROVED** [review]
Approved with nitpicks

File: src/System.Management.Automation/engine/LanguagePrimitives.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/LanguagePrimitives.cs]: They are from reverting those 2 PRs.

File: src/System.Management.Automation/engine/LanguagePrimitives.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/LanguagePrimitives.cs]: Ditto.

File: src/System.Management.Automation/engine/LanguagePrimitives.cs (hunk 3 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/LanguagePrimitives.cs]: Ditto. It's just `git revert <commit-id>`

File: src/System.Management.Automation/engine/LanguagePrimitives.cs (hunk 5 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/LanguagePrimitives.cs]: Yes, it's from `git revert`

**Comment by daxian-dbw** [comment]
@adityapatwardhan You just need to review the 3rd commit. The previous 2 are from `git revert` those 2 PRs.

File: src/System.Management.Automation/engine/LanguagePrimitives.cs (hunk 4 above)
Comment (PoshChan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs]: @daxian-dbw, your last commit had 2 failures in `PowerShell-CI-linux`
Built-in type information tests.Should have correct number of built-in type items in type table
```powershell
Expected exactly 273, but got 271.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/ETS/TypeTable.Tests.ps1: line 22
22:         $types.Count | Should -BeExactly ($isUnixStatEnabled ? 272 : 273)
```
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Diagnostics.TraceSource.dll; SideIndicator=<=}, @{InputObject=System.Runtime.Serialization.Primitives.dll; SideIndicator=<=}).
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 125
125:         $diffs | Should -BeExactly $null
```

Comment (PoshChan) [inline src/System.Management.Automation/engine/LanguagePrimitives.cs]: @daxian-dbw, your last commit had 2 failures in `PowerShell-CI-macos`
Built-in type information tests.Should have correct number of built-in type items in type table
```powershell
Expected exactly 273, but got 271.
at <ScriptBlock>, /Users/runner/runners/2.163.1/work/1/s/test/powershell/engine/ETS/TypeTable.Tests.ps1: line 22
22:         $types.Count | Should -BeExactly ($isUnixStatEnabled ? 272 : 273)
```
Validate start of console host.No new assemblies are loaded
```powershell
Expected exactly $null, but got @(@{InputObject=System.Diagnostics.TraceSource.dll; SideIndicator=<=}, @{InputObject=System.Runtime.Serialization.Primitives.dll; SideIndicator=<=}).
at <ScriptBlock>, /Users/runner/runners/2.163.1/work/1/s/test/powershell/Host/Startup.Tests.ps1: line 125
125:         $diffs | Should -BeExactly $null
```

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/LanguagePrimitives.cs]: Cherry-picked https://github.com/PowerShell/PowerShell/pull/11503 to the release branch.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/LanguagePrimitives.cs]: @adityapatwardhan Can you please take a look at those 2 failures?
The first failure is from the test change in https://github.com/PowerShell/PowerShell/pull/11370. The changes in that PR seem not right when experimental features are disabled by default.
For the second, I remember seeing it somewhere, but couldn't remember now. Do you remember seeing it?

**Comment by daxian-dbw** [comment]
@adityapatwardhan @TravisEz13 Also, the windows build failed at the packaging phase with a "calling method on null-value" error:
```
You cannot call a method on a null-valued expression.
at Invoke-CIFinish, D:\a\1\s\tools\ci.psm1: line 452
at <ScriptBlock>, D:\a\_temp\00413339-7785-4f63-ab8a-bab4a33c6cbe.ps1: line 4
at <ScriptBlock>, <No file>: line 1
You cannot call a method on a null-valued expression.
At D:\a\1\s\tools\ci.psm1:452 char:9
+         $previewLabel = $previewVersion[1].replace('.','')
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
```
See https://dev.azure.com/powershell/PowerShell/_build/results?buildId=43557&view=logs&j=1057fe42-c88b-5deb-88ad-59245c714665&t=1888043b-254d-5761-80ec-47d0d8adc59a

File: src/System.Management.Automation/engine/LanguagePrimitives.cs (hunk 4 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/LanguagePrimitives.cs]: For the second, I remember now we saw it in the release test run for the rc.1 release. We didn't think it was a blocker as it's OK to load less assemblies. But I don't remember if we find out what experimental feature caused this difference.

**Comment by adityapatwardhan** [comment]
@SteveL-MSFT  The PR is targetting RC.2 branch so it is auto approved for RC.

Merged this even though there were failures in Windows packaging as we have differences in content in metadata.json in RC.2 branch. It was fixed to unblock CI with PR #11372 
We do not need to take this change to the RC.2 branch.

**Comment by ili101** [comment]
Why was this reverted?
Is this final or are you planing to change it again in the future?
BTW probably should have been listed under the "Breaking Changes" category in the v7.0.0-rc.2 Release note.

**Comment by vexx32** [comment]
@ili101 the change was never included in a stable release, so reverting it is not a breaking change. Please see [this issue](https://github.com/PowerShell/PowerShell/issues/11604) for discussion -- depending on what is decided by the PS team and community, the change may be reintroduced, likely with some alterations if it is, in a later release, likely at least 7.1 🙁
